File: python project/VectorCollection.py
# Contains a collection of vectors (Documents or Queries) and preprocess operations
# Eric LaBouve (elabouve@calpoly.edu)

from enum import Enum
from DocumentVector import DocumentVector
from QueryVector import QueryVector
from Posting import Posting
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from collections import defaultdict
from WordNet import WordNet
import DistanceFunctions as dist_fs
import re, sys, nltk


class VectorType(Enum):
    DOCUMENTS = 1
    QUERIES = 2


class VectorCollection:
    # Set of stop words
    #stop_words = set(stopwords.words('english'))

    stop_words = {"a", "about", "above", "all", "along",
            "also", "although", "am", "an", "and", "any", "are", "aren't", "as", "at",
            "be", "because", "been", "but", "by", "can", "cannot", "could", "couldn't",
            "did", "didn't", "do", "does", "doesn't", "e.g.", "either", "etc", "etc.",
            "even", "ever", "enough", "for", "from", "further", "get", "gets", "got", "had", "have",
            "hardly", "has", "hasn't", "having", "he", "hence", "her", "here",
            "hereby", "herein", "hereof", "hereon", "hereto", "herewith", "him",
            "his", "how", "however", "i", "i.e.", "if", "in", "into", "it", "it's", "its",
            "me", "more", "most", "mr", "my", "near", "nor", "now", "no", "not", "or", "on", "of", "onto",
            "other", "our", "out", "over", "really", "said", "same", "she",
            "should", "shouldn't", "since", "so", "some", "such",
            "than", "that", "the", "their", "them", "then", "there", "thereby",
            "therefore", "therefrom", "therein", "thereof", "thereon", "thereto",
            "therewith", "these", "they", "this", "those", "through", "thus", "to",
            "too", "under", "until", "unto", "upon", "us", "very", "was", "wasn't",
            "we", "were", "what", "when", "where", "whereby", "wherein", "whether",
            "which", "while", "who", "whom", "whose", "why", "with", "without",
            "would", "you", "your", "yours", "yes"}

    def __init__(self, file_path, vector_type, stop_words_on=False, stemming_on=False, min_word_len=2):
        # Maps {doc id : TextVector}
        self.id_to_textvector = {}
        # Maps {string term : {doc id : Posting}}
        self.term_to_postings = defaultdict(dict)
        # Initialize Stemmer
        self.stemmer = PorterStemmer()
        # Initialize WordNet
        self.wn = WordNet()
        # Parse Documents or Queries
        if vector_type == VectorType.DOCUMENTS:
            self.parse_documents(file_path, stop_words_on, stemming_on, min_word_len)
        elif vector_type == VectorType.QUERIES:
            self.parse_queries(file_path, stop_words_on, stemming_on, min_word_len)

    def __repr__(self):
        s = 'id_to_textvector:\n\tKEY\t\tVALUE\n'
        a = 0
        for key, value in self.id_to_textvector.items():
            s += '\t' + str(key) + '\t\t' + str(value) + '\n'
            if a == 5:
                break
            a += 1
        s += 'term_to_postings:\n\tKEY\t\tVALUE\n'
        a = 0
        for key, value in self.term_to_postings.items():
            s += '\t' + str(key) + '\t\t' + str(value) + '\n'
            if a == 5:
                break
            a += 1
        return s

# __________________Read VectorCollection File Methods__________________

    # term_offset - absolute position of term inside the document
    # paragraph - paragraph number of term inside the document
    # sentence - sentence number of term inside the document
    # word - word number of term inside the document
    def add_to_inverted_index(self, doc_id, term, term_offset, sentence):
        term_postings = self.term_to_postings[term] # {doc id : Posting}
        if term_postings.get(doc_id) is None:
            p = Posting()
            p.add_doc_id(doc_id)
            term_postings[doc_id] = p
        term_postings[doc_id].add_offset(term_offset)
        term_postings[doc_id].add_sentence(sentence)

    # Fill out inverted index and term frequency table
    # Called from inside parse_documents and parse_queries
    def evaluate_vectors(self, min_word_len: int, stemming_on: bool, stop_words_on: bool):
        # For each text vector
        counter = 0
        for textvector_id, textvector in self.id_to_textvector.items():
            counter += 1
            if counter % 500 == 0:
                print("Vector " + str(counter) + " Processed", end=", ")
            cur_offset = 0  # Offset of the term inside the text vector
            sentence_num = 0
            next_sentence = False
            try:
                tagged_text = self.tag_text(textvector_id)
                # Loop through each vector term
                for term, pos in tagged_text:
                    if '.' in term:
                        next_sentence = True
                    term = re.sub(r'[^\w\s]', '', term.lower())  # Remove punctuation
                    if len(term) >= min_word_len:  # Satisfies min length
                        # Stem all terms
                        if stemming_on:
                            stem_term = self.stemmer.stem(term)
                        # If (stop words not on and term is not a stop word) or (stop words on)
                        if ((not stop_words_on) and (term not in self.stop_words)) or stop_words_on:
                            # Expand if vector is a query
                            self.expand_query(stemming_on, term, textvector)
                            textvector.add_term(stem_term)  # Add term to term_to_freq
                            textvector.terms.append(stem_term)
                            textvector.terms_pos.append(pos)
                            # Add term to inverted index
                            self.add_to_inverted_index(textvector_id, stem_term, cur_offset, sentence_num)
                            cur_offset += 1
                    if next_sentence:
                        next_sentence = False
                        sentence_num += 1
            except IndexError:
                print("Index Error at evaluate_vectors in VectorCollection")

    # WordNet query term expansion
    def expand_query(self, stemming_on, term, textvector):
        # Include word substitutions if textvector is a query
        if type(textvector) == QueryVector:
            subs = self.wn.get_sim_terms_rw(term)
            if stemming_on:
                textvector.terms_sub.append(self.wn.stem(self.stemmer, term, subs))
            else:
                textvector.terms_sub.append(subs)

    def tag_text(self, textvector_id):
        split_text = re.split("[^a-zA-Z.]+", self.id_to_textvector[textvector_id].raw_text)
        if len(split_text) > 0 and split_text[-1] == '':
            del split_text[-1]
        if len(split_text) > 0 and split_text[0] == '':
            del split_text[0]
        tagged_text = nltk.pos_tag(split_text)
        return tagged_text

    def parse_documents(self, file_path: str, stop_words_on: bool, stemming_on: bool, min_word_len: int):
        # Read vectors into memory
        with open(file_path) as file:
            cur_doc_id = -1  # Current Doc Id
            inside_W = False
            for line in file:
                if '.I' in line: # Contains Id
                    cur_doc_id = int(re.sub(r'\s+', '', re.sub(r".I", '', line)))  # Remove .I and spaces
                    textvector = DocumentVector()
                    textvector.add_id(cur_doc_id)
                    self.id_to_textvector[cur_doc_id] = textvector
                    inside_W = False
                elif '.W' in line:
                    inside_W = True
                elif inside_W:  # In the body of the Document
                    self.id_to_textvector[cur_doc_id].raw_text += re.sub('\n', ' ', line)

        print("Building Document Vectors")
        # Fill out inverted index and term frequency table
        self.evaluate_vectors(min_word_len, stemming_on, stop_words_on)

    def parse_queries(self, file_path: str, stop_words_on: bool, stemming_on: bool, min_word_len: int):
        # Read vectors into memory
        with open(file_path) as file:
            cur_query_id = 0  # Current Query Id
            inside_W = False
            for line in file:
                if '.I' in line:  # Contains Id
                    cur_query_id += 1
                    textvector = QueryVector()
                    textvector.add_id(cur_query_id)
                    self.id_to_textvector[cur_query_id] = textvector
                    inside_W = False
                elif '.W' in line:
                    inside_W = True
                elif inside_W:  # In the body of the Document
                    self.id_to_textvector[cur_query_id].raw_text += re.sub('\n', ' ', line)

        print("Building Query Vectors")
        # Fill out inverted index and term frequency table
        self.evaluate_vectors(min_word_len, stemming_on, stop_words_on)

# __________________Normalization Methods__________________

    # Normalize the current VectorCollection against the vectors in the given VectorCollection
    # vector_collection is intended to be the document textvector collection
    def normalize(self, vector_collection):
        for id, textvector in self.id_to_textvector.items():
            textvector.normalize(vector_collection)

    # Returns the number of vectors in the collection
    def get_num_vectors(self):
        return len(self.id_to_textvector)

# __________________Distance Methods__________________

    # Computes ALL distances for ALL Queries x Documents.
    # Stores results in a map {Query id : [Doc Ids]} where Doc Ids are
    # sorted in order of relevance.
    # self - Intended to be the Query VectorCollection
    # documents - Intended to be the Documents VectorCollection
    # dist_obj - A object that holds a distance function
    # doc_limit - An upper limit for the number of document ids returned per query. If negative, use all documents
    # query_limit - An upper limit for the number of queries to process. If negative, use all queries
    # returns a map {Query id : [Doc Ids]}
    def find_closest_docs(self, documents, dist_obj, doc_limit=-1, query_limit=-1) -> map:
        results = {}
        computed = 0
        for qry_id, qry_vector in self.id_to_textvector.items():
            dist_obj.set_query(qry_vector)
            ranked_docs = dist_fs.find_closest_docs(documents, dist_obj, doc_limit=doc_limit)
            results[qry_id] = ranked_docs
            computed += 1
            sys.stdout.write("Q:" + str(qry_id) + " ")
            sys.stdout.flush()
            if computed == query_limit:
                break
        return results

# __________________Inverted Index Methods__________________

    # Returns the number of vectors that contains the term
    def get_doc_freq(self, term):
        if term in self.term_to_postings:
            return len(self.term_to_postings[term])
        return 0

    # Returns a Posting for a specified term and document
    def get_term_posting_for_doc(self, term: str, doc_id: int) -> Posting:
        # Check if term exists in inverted index
        if term in self.term_to_postings:
            doc_postings = self.term_to_postings[term]
            # Check if document contains term
            if doc_id in doc_postings:
                return doc_postings[doc_id]
        return None

File: java project/src/DocumentClasses/VectorCollection.java
package DocumentClasses;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.Serializable;
import java.util.*;

/**
 * Holds a mapping from IDs to TextVectors
 *
 * @author ericlabouve
 */
public class VectorCollection implements Serializable {

    /** Maps each document ID to a TextVector*/
    private HashMap<Integer, TextVector> idToTextVectorMap;

    /** Inverted Index - A mapping from terms to postings
     * Stores which terms are stored in which documents.
     * Also stores where in the document the term is stored. */
    private HashMap<String, Posting> termToPostingMap;

    /** Whether or not to skip over noise words. */
    private boolean noiseWordsOn;

    /** The minumum character count for a word to be included in the text vector */
    private int minWordLength;

    /** Set of all noise words represented in noiseWordArray */
    private Set<String> noiseWordsSet = null;
    private static  String noiseWordArray[] = {"a", "about", "above", "all", "along",
            "also", "although", "am", "an", "and", "any", "are", "aren't", "as", "at",
            "be", "because", "been", "but", "by", "can", "cannot", "could", "couldn't",
            "did", "didn't", "do", "does", "doesn't", "e.g.", "either", "etc", "etc.",
            "even", "ever", "enough", "for", "from", "further", "get", "gets", "got", "had", "have",
            "hardly", "has", "hasn't", "having", "he", "hence", "her", "here",
            "hereby", "herein", "hereof", "hereon", "hereto", "herewith", "him",
            "his", "how", "however", "i", "i.e.", "if", "in", "into", "it", "it's", "its",
            "me", "more", "most", "mr", "my", "near", "nor", "now", "no", "not", "or", "on", "of", "onto",
            "other", "our", "out", "over", "really", "said", "same", "she",
            "should", "shouldn't", "since", "so", "some", "such",
            "than", "that", "the", "their", "them", "then", "there", "thereby",
            "therefore", "therefrom", "therein", "thereof", "thereon", "thereto",
            "therewith", "these", "they", "this", "those", "through", "thus", "to",
            "too", "under", "until", "unto", "upon", "us", "very", "was", "wasn't",
            "we", "were", "what", "when", "where", "whereby", "wherein", "whether",
            "which", "while", "who", "whom", "whose", "why", "with", "without",
            "would", "you", "your", "yours", "yes"};

    public enum VectorType {
        DOCUMENTS, QUIRIES
    }

    /**
     * Reads the file that is specified as input to populate |idToTextVectorMap|.
     * @param filePath path to an input vectors file.
     * @param vectorType If equal to DOCUMENTS, adds DocumentVector objects to HashMap.
     * @param noiseWordsOn Whether or not to skip over noise words.
     * @param minWordLength The minumum character count for a word to be included in the text vector
     */
    public VectorCollection(String filePath, VectorType vectorType, boolean noiseWordsOn, int minWordLength) {
        this.noiseWordsOn = noiseWordsOn;
        this.minWordLength = minWordLength;
        idToTextVectorMap = new HashMap<Integer, TextVector>();
        termToPostingMap = new HashMap<String, Posting>();
        if (vectorType == VectorType.DOCUMENTS) {
            readDocumentVectors(filePath, noiseWordsOn, minWordLength);
        } else if (vectorType == VectorType.QUIRIES) {
            readQueryVectors(filePath, noiseWordsOn, minWordLength);
        }
    }

    /**
     * Parses the text indicated by |filePath| and populates the idToTextVectorMap with
     * IDs that map to document vectors and the termToPostingMap which is an inverted index
     * that maps terms to documentIds+locations
     *
     * Tag Descriptions:
     * .I - ID
     * .T - Title
     * .A - Author(s)
     * .B - Publisher
     * .W - Text body
     * @param filePath path to a Document vectors file.
     */
    private void readDocumentVectors(String filePath, boolean noiseWordsOn, int minWordLength) {
        try(BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            int curDocId = -1; // Current Document ID
            int curDocIdx = 0; // Current Document Index
            boolean insideW = false;
            String line = br.readLine();
            // Continue until end of file is reached
            while (line != null) {
                // Check if we have reached a new document
                if (line.contains(".I")) {
                    // The current Document ID is the number after .I
                    curDocId = Integer.valueOf(line.replace(".I", "").replaceAll("\\s+",""));
                    TextVector tVector = new DocumentVector();
                    idToTextVectorMap.put(curDocId, tVector);
                    tVector.setId(curDocId);
                    insideW = false;
                }
                // If next line will indicate that we are parsing the text body
                else if (line.contains(".W")) {
                    insideW = true;
                    curDocIdx = 0; // Reset for next document
                }
                // If we are about to parse the text body
                else if (insideW) {
                    // Split into individual words
                    String[] tokens = line.split("[^a-zA-Z]+");
                    for (int i = 0; i < tokens.length; i++) {
                        String term = tokens[i].toLowerCase();
                        // Store words that are appropriate length
                        if (term.length() >= minWordLength) {

                            // If noise words are off and the current word is not a noise word
                            if(!noiseWordsOn && !isNoiseWord(term)) {
                                // Add to text vector
                                getVectorById(curDocId).add(term);
                                // Add to inverted index
                                addTermToPostingMap(term, curDocId, curDocIdx);
                                curDocIdx++;
                            }

                            // Add any word if noise words are on
                            else if (noiseWordsOn) {
                                // Add to text vector
                                getVectorById(curDocId).add(term);
                                // Add to inverted index
                                addTermToPostingMap(term, curDocId, curDocIdx);
                                curDocIdx++;
                            }
                        }
                    }
                }
                line = br.readLine();
            }
        } catch (java.io.FileNotFoundException ex) {
            ex.printStackTrace();
        } catch (java.io.IOException ex) {
            ex.printStackTrace();
        }
    }

    /**
     * Parses the text indicated by |filePath| and populates idToTextVectorMap with
     * IDs that map to document vectors
     *
     * Tag Descriptions:
     * .I - ID
     * .W - Text body
     * @param filePath Path to the Query vector file
     */
    public void readQueryVectors(String filePath, boolean noiseWordsOn, int minWordLength) {
        try(BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            int curQuId = 0; //  Current Query ID
            int curQuIdx = 0; // Current Document Index
            boolean insideW = false;
            String line = br.readLine();
            // Continue until end of file is reached
            while (line != null) {
                // Check if we have reached a new document
                if (line.contains(".I")) {
                    // The current Document is the number after .I
                    curQuId++;
                    TextVector tVector = new QueryVector();
                    idToTextVectorMap.put(curQuId, tVector);
                    tVector.setId(curQuId);
                    insideW = false;
                }
                else if (line.contains(".W")) {
                    insideW = true;
                    curQuIdx = 0; // Reset for next query
                }
                else if (insideW) {
                    String[] tokens = line.split("[^a-zA-Z]+");
                    for (int i = 0; i < tokens.length; i++) {
                        String term = tokens[i].toLowerCase();
                        // Store words that are appropriate length
                        if (term.length() >= minWordLength) {
                            // If noise words are off and the current word is not a noise word
                            if(!noiseWordsOn && !isNoiseWord(tokens[i])) {
                                // Add to text vector
                                getVectorById(curQuId).add(term);
                                // Add to query term to rawText
                                getVectorById(curQuId).appendToRawText(term);
                                // Add to inverted index
                                addTermToPostingMap(term, curQuId, curQuIdx);
                                curQuIdx++;
                            }
                            // Add any word if noise words are on
                            else if (noiseWordsOn) {
                                // Add to text vector
                                getVectorById(curQuId).add(term);
                                // Add to query term to rawText
                                getVectorById(curQuId).appendToRawText(term);
                                // Add to inverted index
                                addTermToPostingMap(term, curQuId, curQuIdx);
                                curQuIdx++;
                            }
                        }
                    }
                }
                line = br.readLine();
            }
        } catch (java.io.FileNotFoundException ex) {
            ex.printStackTrace();
        } catch (java.io.IOException ex) {
            ex.printStackTrace();
        }
    }

/****************************************** Inverted Index Methods ******************************************/

    /**
     * Adds a term to the termToPostingMap with a particular document id and index
     * @param term The word term to insert into the termToPostingMap
     * @param docId The document id for the term
     * @param idx The index into the document given by docId for a particular term
     */
    private void addTermToPostingMap(String term, int docId, int idx) {
        if (!termToPostingMap.containsKey(term))
            termToPostingMap.put(term, new Posting());
        termToPostingMap.get(term).add(docId, idx);
    }

    public Posting getPosting(String term) {
        return termToPostingMap.get(term.toLowerCase());
    }

    public boolean hasPosting(String term) {
        return termToPostingMap.get(term.toLowerCase()) != null;
    }

/****************************************** Other Methods ******************************************/

    public boolean areNoiseWordsOn() {
        return noiseWordsOn;
    }

    public int getMinWordLength() {
        return minWordLength;
    }

    /**
     * Normalizes all the idToTextVectorMap in this collection of idToTextVectorMap with
     * w = tf * idf
     * calls the normalize() method on each document in the collection
     *
     * When you normalize, you pass in the idToTextVectorMap because you care about how many DOCUMENTS contain our search word
     */
    public void normalize(VectorCollection documents) {
        for (TextVector vector : getTextVectors()) {
            vector.normalize(documents);
        }
    }

    /**
     * returns the TextVector for the document with the ID that is given.
     */
    public TextVector getVectorById(int id) {
        return idToTextVectorMap.get(id);
    }

    /**
     * @return the average length of a vector.
     */
    public double getAverageVectorLength() {
        int runningTotal = 0;
        for (TextVector tVector : idToTextVectorMap.values()) {
            runningTotal += tVector.getTotalWordCount();
        }
        return runningTotal / (double) idToTextVectorMap.size();
    }

    /**
     * @return number of idToTextVectorMap
     */
    public int getSize() {
        return idToTextVectorMap.size();
    }

    /**
     * @return a Collection of TextVectors
     */
    public Collection<TextVector> getTextVectors() {
        return idToTextVectorMap.values();
    }

    /**
     * @return a mapping of document id to Text Vector, that is an object of type Set<Map.Entry<Integer, TextVector>>.
     */
    public Set<Map.Entry<Integer, TextVector>> getEntrySet() {
        return idToTextVectorMap.entrySet();
    }

    /**
     * Computes the document frequency for an input word
     * @return the number of vectors in the idToTextVectorMap collection that contains the input word
     */
    public int getDocumentFrequency(String inputWord) {
        int freq = 0;
        for (TextVector doc : idToTextVectorMap.values()) {
            if (doc.contains(inputWord)) {
                freq++;
            }
        }
        return freq;
    }

    /**
     * is the input a noise word.
     */
    private boolean isNoiseWord(String word) {
        // Lazy loading
        if (noiseWordsSet == null) {
            noiseWordsSet = new HashSet<String>(Arrays.asList(VectorCollection.noiseWordArray));
        }
        return noiseWordsSet.contains(word);
    }
}
